server/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java

```diff
@@ -57,60 +68,145 @@ private static XContentType getResponseContentType(RestRequest request) {
             }
-            out.append("\n");
-        }
 
-        List<Integer> rowOrder = getRowOrder(table, request);
+            @Override
+            public ReleasableBytesReference encodeChunk(int sizeHint, Recycler<BytesRef> recycler) throws IOException {
+                try {
+                    assert currentOutput == null;
+                    currentOutput = new RecyclerBytesStreamOutput(recycler);
+
+                    if (needsHeader) {
+                        needsHeader = false;
+                        for (int col = 0; col < headers.size(); col++) {
+                            DisplayHeader header = headers.get(col);
+                            boolean isLastColumn = col == lastHeader;
+                            pad(
+                                new Table.Cell(header.display, table.findHeaderByName(header.name)),
+                                width[col],
+                                request,
+                                writer,
+                                isLastColumn
+                            );
+                            if (isLastColumn == false) {
+                                writer.append(" ");
+                            }
+                        }
+                        writer.append("\n");
+                    }
 
-        for (Integer row : rowOrder) {
-            for (int col = 0; col < headers.size(); col++) {
-                DisplayHeader header = headers.get(col);
-                boolean isLastColumn = col == lastHeader;
-                pad(table.getAsMap().get(header.name).get(row), width[col], request, out, isLastColumn);
-                if (isLastColumn == false) {
-                    out.append(" ");
+                    while (rowIterator.hasNext() && currentOutput.size() < sizeHint) {
+                        final var row = rowIterator.next();
+                        for (int col = 0; col < headers.size(); col++) {
+                            DisplayHeader header = headers.get(col);
+                            boolean isLastColumn = col == lastHeader;
+                            pad(table.getAsMap().get(header.name).get(row), width[col], request, writer, isLastColumn);
+                            if (isLastColumn == false) {
+                                writer.append(" ");
+                            }
+                        }
+                        writer.append("\n");
+                    }
+
+                    if (rowIterator.hasNext()) {
+                        writer.flush();
+                    } else {
+                        writer.close();
+                    }
+
+                    final var chunkOutput = currentOutput;
+                    currentOutput = null;
+                    return new ReleasableBytesReference(chunkOutput.bytes(), () -> Releasables.closeExpectNoException(chunkOutput));
+                } finally {
+                    if (currentOutput != null) {
```

Comment:
Although I don't see any obvious ways to throw exceptions in this area, I'm somewhat dubious about how such an exception would be handled in the networking code.